url/url.cpp

```diff
@@ -51,4 +101,1341 @@ std::string blob_url_create(Origin const &origin) {
+
+                    if (!ipv4_piece.has_value()) {
+                        ipv4_piece = number;
+                    } else if (*ipv4_piece == 0) {
+                        validation_error(ValidationError::IPv4InIPv6InvalidCodePoint);
+
+                        return std::nullopt;
+                    } else {
+                        ipv4_piece = *ipv4_piece * 10 + number;
+                    }
+
+                    if (*ipv4_piece > 255) {
+                        validation_error(ValidationError::IPv4InIPv6OutOfRangePart);
+
+                        return std::nullopt;
+                    }
+
+                    pointer++;
+                }
+
+                address[piece_index] = static_cast<std::uint16_t>(address[piece_index] * 0x100ul + *ipv4_piece);
+
+                numbers_seen++;
+
+                if (numbers_seen == 2 || numbers_seen == 4) {
+                    piece_index++;
+                }
+            }
+
+            if (numbers_seen != 4) {
+                validation_error(ValidationError::IPv4InIPv6TooFewParts);
+
+                return std::nullopt;
+            }
+
+            break;
+        } else if (input[pointer] == ':') {
+            pointer++;
+
+            if (pointer >= input.size()) {
+                validation_error(ValidationError::IPv6InvalidCodePoint);
+
+                return std::nullopt;
+            }
+        } else if (pointer < input.size()) {
+            validation_error(ValidationError::IPv6InvalidCodePoint);
+
+            return std::nullopt;
+        }
+
+        address[piece_index] = static_cast<std::uint16_t>(value);
+
+        piece_index++;
+    }
+
+    if (compress.has_value()) {
+        std::size_t swaps = piece_index - *compress;
+
+        piece_index = 7;
+
+        for (; piece_index != 0 && swaps > 0; piece_index--, swaps--) {
+            std::uint16_t tmp = address[piece_index];
+            address[piece_index] = address[*compress + swaps - 1];
+            address[*compress + swaps - 1] = tmp;
+        }
+    } else if (!compress.has_value() && piece_index != 8) {
+        validation_error(ValidationError::IPv6TooFewPieces);
+
+        return std::nullopt;
+    }
+
+    return address;
+}
+
+bool UrlParser::is_url_codepoint(std::uint32_t cp) {
+    return cp == '!' || cp == '$' || cp == '&' || cp == '\'' || cp == '(' || cp == ')' || cp == '*' || cp == '+'
+            || cp == ',' || cp == '-' || cp == '.' || cp == '/' || cp == ':' || cp == ';' || cp == '=' || cp == '?'
+            || cp == '@' || cp == '_' || cp == '~'
+            || (cp >= 0x00a0 && cp <= 0x10fffd && !util::codepoint_is_noncharacter(cp)
+                    && !util::codepoint_is_surrogate(cp));
+}
+
+// https://url.spec.whatwg.org/#concept-opaque-host-parser
+std::optional<std::string> UrlParser::parse_opaque_host(std::string_view input) {
+    if (input.find_first_of("\0\t\n\r #/:<>?@[\\]^|", 0, 17) != std::string_view::npos) {
```

Comment:
Only looking at the first 17 characters of input seems unintentional? Just drop the last 2 args to `std::string_view::find_first_of` if you want to search the full string.